In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [2]:
url1 = "datasets/books.csv"
url2 = "datasets/ratings.csv"
url3 = "datasets/users.csv"

books = pd.read_csv(url1, encoding='latin-1')
ratings = pd.read_csv(url2, encoding='latin-1')
users = pd.read_csv(url3, encoding='latin-1')
books.head()

C:\Users\sadma\AppData\Local\Temp\ipykernel_16112\2952862037.py:5: DtypeWarning: Columns (3) have mixed types. Specify dtype option on import or set low_memory=False.
  books = pd.read_csv(url1, encoding='latin-1')


ISBN                                         Book-Title  \
0  0195153448                                Classical Mythology   
1  0002005018                                       Clara Callan   
2  0060973129                               Decision in Normandy   
3  0374157065  Flu: The Story of the Great Influenza Pandemic...   
4  0393045218                             The Mummies of Urumchi   

            Book-Author Year-Of-Publication                   Publisher  \
0    Mark P. O. Morford                2002     Oxford University Press   
1  Richard Bruce Wright                2001       HarperFlamingo Canada   
2          Carlo D'Este                1991             HarperPerennial   
3      Gina Bari Kolata                1999        Farrar Straus Giroux   
4       E. J. W. Barber                1999  W. W. Norton &amp; Company   

                                         Image-URL-S  \
0  http://images.amazon.com/images/P/0195153448.0...   
1  http://images.amazon.com/images/P/0002005018.0...   
2  http://images.amazon.com/images/P/0060973129.0...   
3  http://images.amazon.com/images/P/0374157065.0...   
4  http://images.amazon.com/images/P/0393045218.0...   

                                         Image-URL-M  \
0  http://images.amazon.com/images/P/0195153448.0...   
1  http://images.amazon.com/images/P/0002005018.0...   
2  http://images.amazon.com/images/P/0060973129.0...   
3  http://images.amazon.com/images/P/0374157065.0...   
4  http://images.amazon.com/images/P/0393045218.0...   

                                         Image-URL-L  
0  http://images.amazon.com/images/P/0195153448.0...  
1  http://images.amazon.com/images/P/0002005018.0...  
2  http://images.amazon.com/images/P/0060973129.0...  
3  http://images.amazon.com/images/P/0374157065.0...  
4  http://images.amazon.com/images/P/0393045218.0...

In [3]:
ratings.head()

User-ID        ISBN  Book-Rating
0   276725  034545104X            0
1   276726  0155061224            5
2   276727  0446520802            0
3   276729  052165615X            3
4   276729  0521795028            6

In [4]:
users.head()

User-ID                            Location   Age
0        1                  nyc, new york, usa   NaN
1        2           stockton, california, usa  18.0
2        3     moscow, yukon territory, russia   NaN
3        4           porto, v.n.gaia, portugal  17.0
4        5  farnborough, hants, united kingdom   NaN

In [5]:
books.columns = books.columns.str.lower().str.replace(" ", "_").str.replace("-", "_")
ratings.columns = ratings.columns.str.lower().str.replace(" ", "_").str.replace("-", "_")
users.columns = users.columns.str.lower().str.replace(" ", "_").str.replace("-", "_")

In [6]:
books.rename(columns={
    "book_title": "book_name",
    "book_author": "author",
    "image_url_l": "image_url"
    }, inplace=True)
books.columns

Index(['isbn', 'book_name', 'author', 'year_of_publication', 'publisher',
       'image_url_s', 'image_url_m', 'image_url'],
      dtype='object')

In [7]:
books.drop([
    "image_url_s",
    "image_url_m",
    "year_of_publication"
    ], axis=1, inplace=True)
books.columns

Index(['isbn', 'book_name', 'author', 'publisher', 'image_url'], dtype='object')

In [8]:
ratings.rename(columns={
    "book_rating": "ratings"
    }, inplace=True)
ratings.columns


Index(['user_id', 'isbn', 'ratings'], dtype='object')

In [9]:
users.drop(columns={
    "age"
    }, inplace=True)
users.columns

Index(['user_id', 'location'], dtype='object')

In [10]:
print(books.shape)
print(ratings.shape)
print(users.shape)

(271360, 5)
(1149780, 3)
(278858, 2)


#### There are 271360 books 
#### There are 278858 users
#### 278858 users rated 1149780

In [11]:
books.isnull().sum()

isbn         0
book_name    0
author       2
publisher    2
image_url    3
dtype: int64

In [12]:
books.dropna(inplace=True)

In [13]:
books.isnull().sum()

isbn         0
book_name    0
author       0
publisher    0
image_url    0
dtype: int64

In [14]:
ratings.isnull().sum()

user_id    0
isbn       0
ratings    0
dtype: int64

In [15]:
users.isnull().sum()

user_id     0
location    0
dtype: int64

In [16]:
books.duplicated().sum()

np.int64(0)

In [17]:
ratings.duplicated().sum()

np.int64(0)

In [18]:
users.duplicated().sum()

np.int64(0)

## Step 1: Merge Rarings and Books Together

In [19]:
ratings_with_books = ratings.merge(books, on="isbn", how="inner")
ratings_with_books.head()

user_id        isbn  ratings  \
0   276725  034545104X        0   
1   276726  0155061224        5   
2   276727  0446520802        0   
3   276729  052165615X        3   
4   276729  0521795028        6   

                                           book_name           author  \
0                               Flesh Tones: A Novel       M. J. Rose   
1                                   Rites of Passage       Judith Rae   
2                                       The Notebook  Nicholas Sparks   
3                                     Help!: Level 1    Philip Prowse   
4  The Amsterdam Connection : Level 4 (Cambridge ...      Sue Leather   

                    publisher  \
0            Ballantine Books   
1                      Heinle   
2                Warner Books   
3  Cambridge University Press   
4  Cambridge University Press   

                                           image_url  
0  http://images.amazon.com/images/P/034545104X.0...  
1  http://images.amazon.com/images/P/0155061224.0...  
2  http://images.amazon.com/images/P/0446520802.0...  
3  http://images.amazon.com/images/P/052165615X.0...  
4  http://images.amazon.com/images/P/0521795028.0...

In [20]:
ratings_with_books.shape

(1031128, 7)

## Step 2: find users who rated > 200 books


In [21]:
ratings_with_books.groupby("user_id").count()

isbn  ratings  book_name  author  publisher  image_url
user_id                                                        
2           1        1          1       1          1          1
8          17       17         17      17         17         17
9           3        3          3       3          3          3
10          1        1          1       1          1          1
12          1        1          1       1          1          1
...       ...      ...        ...     ...        ...        ...
278846      1        1          1       1          1          1
278849      4        4          4       4          4          4
278851     23       23         23      23         23         23
278852      1        1          1       1          1          1
278854      8        8          8       8          8          8

[92106 rows x 6 columns]

In [22]:
ratings_with_books.groupby("user_id").count()["ratings"]

user_id
2          1
8         17
9          3
10         1
12         1
          ..
278846     1
278849     4
278851    23
278852     1
278854     8
Name: ratings, Length: 92106, dtype: int64

In [23]:
ratings_with_books.groupby("user_id").count()["ratings"] > 200

user_id
2         False
8         False
9         False
10        False
12        False
          ...  
278846    False
278849    False
278851    False
278852    False
278854    False
Name: ratings, Length: 92106, dtype: bool

In [24]:
users_rated_books_more_then_200 = ratings_with_books.groupby("user_id").count()["ratings"] > 200
users_rated_books_more_then_200[users_rated_books_more_then_200]

user_id
254       True
2276      True
2766      True
2977      True
3363      True
          ... 
274308    True
275970    True
277427    True
277639    True
278418    True
Name: ratings, Length: 811, dtype: bool

In [25]:
users_rated_books_more_then_200[users_rated_books_more_then_200].index

Index([   254,   2276,   2766,   2977,   3363,   4017,   4385,   6251,   6323,
         6543,
       ...
       271705, 273979, 274004, 274061, 274301, 274308, 275970, 277427, 277639,
       278418],
      dtype='int64', name='user_id', length=811)

In [26]:
users_rated_books_more_then_200 = users_rated_books_more_then_200[users_rated_books_more_then_200].index
users_rated_books_more_then_200

Index([   254,   2276,   2766,   2977,   3363,   4017,   4385,   6251,   6323,
         6543,
       ...
       271705, 273979, 274004, 274061, 274301, 274308, 275970, 277427, 277639,
       278418],
      dtype='int64', name='user_id', length=811)

In [27]:
ratings_with_books["user_id"].isin(users_rated_books_more_then_200)

0          False
1          False
2          False
3          False
4          False
           ...  
1031123    False
1031124    False
1031125    False
1031126    False
1031127    False
Name: user_id, Length: 1031128, dtype: bool

In [28]:
#filter 1
filter1 = ratings_with_books[ratings_with_books["user_id"].isin(users_rated_books_more_then_200)]
filter1

user_id        isbn  ratings  \
1150      277427  002542730X       10   
1151      277427  0026217457        0   
1152      277427  003008685X        8   
1153      277427  0030615321        0   
1154      277427  0060002050        0   
...          ...         ...      ...   
1029349   275970  1931868123        0   
1029350   275970  3411086211       10   
1029351   275970  3829021860        0   
1029352   275970  4770019572        0   
1029353   275970  9626340762        8   

                                                 book_name  \
1150     Politically Correct Bedtime Stories: Modern Ta...   
1151                    Vegetarian Times Complete Cookbook   
1152                                              Pioneers   
1153      Ask for May, Settle for June (A Doonesbury book)   
1154                     On a Wicked Dawn (Cynster Novels)   
...                                                    ...   
1029349  There's a Porcupine in My Outhouse: Misadventu...   
1029350                                         Die Biene.   
1029351                                     The Penis Book   
1029352                                            Musashi   
1029353  Northanger Abbey (Classic Literature with Clas...   

                                      author  \
1150                       James Finn Garner   
1151                              Lucy  Moll   
1152                   James Fenimore Cooper   
1153                           G. B. Trudeau   
1154                       Stephanie Laurens   
...                                      ...   
1029349                         Mike Tougias   
1029350  Sybil Gr�¤fin Sch�¶nfeldt   
1029351                         Joseph Cohen   
1029352                       Eiji Yoshikawa   
1029353                          Jane Austen   

                                    publisher  \
1150                John Wiley &amp; Sons Inc   
1151                    John Wiley &amp; Sons   
1152                         Thomson Learning   
1153                      Henry Holt &amp; Co   
1154                               Avon Books   
...                                       ...   
1029349                    Capital Books (VA)   
1029350  Bibliographisches Institut, Mannheim   
1029351                              Konemann   
1029352          Kodansha International (JPN)   
1029353                 Naxos Audiobooks Ltd.   

                                                 image_url  
1150     http://images.amazon.com/images/P/002542730X.0...  
1151     http://images.amazon.com/images/P/0026217457.0...  
1152     http://images.amazon.com/images/P/003008685X.0...  
1153     http://images.amazon.com/images/P/0030615321.0...  
1154     http://images.amazon.com/images/P/0060002050.0...  
...                                                    ...  
1029349  http://images.amazon.com/images/P/1931868123.0...  
1029350  http://images.amazon.com/images/P/3411086211.0...  
1029351  http://images.amazon.com/images/P/3829021860.0...  
1029352  http://images.amazon.com/images/P/4770019572.0...  
1029353  http://images.amazon.com/images/P/9626340762.0...  

[474002 rows x 7 columns]

## Step 3: find books rated > 50 

In [29]:
filter1.groupby("book_name").count()

user_id  isbn  ratings  \
book_name                                                                    
 A Light in the Storm: The Civil War Diary of A...        2     2        2   
 Always Have Popsicles                                    1     1        1   
 Apple Magic (The Collector's series)                     1     1        1   
 Beyond IBM: Leadership Marketing and Finance f...        1     1        1   
 Clifford Visita El Hospital (Clifford El Gran ...        1     1        1   
...                                                     ...   ...      ...   
Ã?Â?ber das Fernsehen.                                  2     2        2   
Ã?Â?ber die Pflicht zum Ungehorsam gegen den ...        3     3        3   
Ã?Â?lpiraten.                                           1     1        1   
Ã?Â?stlich der Berge.                                   1     1        1   
Ã?Â?thique en toc                                       1     1        1   

                                                    author  publisher  \
book_name                                                               
 A Light in the Storm: The Civil War Diary of A...       2          2   
 Always Have Popsicles                                   1          1   
 Apple Magic (The Collector's series)                    1          1   
 Beyond IBM: Leadership Marketing and Finance f...       1          1   
 Clifford Visita El Hospital (Clifford El Gran ...       1          1   
...                                                    ...        ...   
Ã?Â?ber das Fernsehen.                                 2          2   
Ã?Â?ber die Pflicht zum Ungehorsam gegen den ...       3          3   
Ã?Â?lpiraten.                                          1          1   
Ã?Â?stlich der Berge.                                  1          1   
Ã?Â?thique en toc                                      1          1   

                                                    image_url  
book_name                                                      
 A Light in the Storm: The Civil War Diary of A...          2  
 Always Have Popsicles                                      1  
 Apple Magic (The Collector's series)                       1  
 Beyond IBM: Leadership Marketing and Finance f...          1  
 Clifford Visita El Hospital (Clifford El Gran ...          1  
...                                                       ...  
Ã?Â?ber das Fernsehen.                                    2  
Ã?Â?ber die Pflicht zum Ungehorsam gegen den ...          3  
Ã?Â?lpiraten.                                             1  
Ã?Â?stlich der Berge.                                     1  
Ã?Â?thique en toc                                         1  

[155841 rows x 6 columns]

In [30]:
filter1.groupby("book_name").count()["ratings"]

book_name
 A Light in the Storm: The Civil War Diary of Amelia Martin, Fenwick Island, Delaware, 1861 (Dear America)    2
 Always Have Popsicles                                                                                        1
 Apple Magic (The Collector's series)                                                                         1
 Beyond IBM: Leadership Marketing and Finance for the 1990s                                                   1
 Clifford Visita El Hospital (Clifford El Gran Perro Colorado)                                                1
                                                                                                             ..
Ã?Â?ber das Fernsehen.                                                                                      2
Ã?Â?ber die Pflicht zum Ungehorsam gegen den Staat.                                                         3
Ã?Â?lpiraten.                                                                               

In [31]:
filter1.groupby("book_name").count()["ratings"] > 50

book_name
 A Light in the Storm: The Civil War Diary of Amelia Martin, Fenwick Island, Delaware, 1861 (Dear America)    False
 Always Have Popsicles                                                                                        False
 Apple Magic (The Collector's series)                                                                         False
 Beyond IBM: Leadership Marketing and Finance for the 1990s                                                   False
 Clifford Visita El Hospital (Clifford El Gran Perro Colorado)                                                False
                                                                                                              ...  
Ã?Â?ber das Fernsehen.                                                                                      False
Ã?Â?ber die Pflicht zum Ungehorsam gegen den Staat.                                                         False
Ã?Â?lpiraten.                                               

In [32]:
filter2 = filter1.groupby("book_name").count()["ratings"] >= 50
filter2[filter2]

book_name
1984                                                                 True
1st to Die: A Novel                                                  True
2nd Chance                                                           True
4 Blondes                                                            True
A Bend in the Road                                                   True
                                                                     ... 
Year of Wonders                                                      True
You Belong To Me                                                     True
Zen and the Art of Motorcycle Maintenance: An Inquiry into Values    True
Zoya                                                                 True
\O\" Is for Outlaw"                                                  True
Name: ratings, Length: 706, dtype: bool

In [33]:
filter2[filter2].index

Index(['1984', '1st to Die: A Novel', '2nd Chance', '4 Blondes',
       'A Bend in the Road', 'A Case of Need',
       'A Child Called \It\": One Child's Courage to Survive"',
       'A Civil Action', 'A Day Late and a Dollar Short', 'A Fine Balance',
       ...
       'Winter Solstice', 'Wish You Well', 'Without Remorse',
       'Wizard and Glass (The Dark Tower, Book 4)', 'Wuthering Heights',
       'Year of Wonders', 'You Belong To Me',
       'Zen and the Art of Motorcycle Maintenance: An Inquiry into Values',
       'Zoya', '\O\" Is for Outlaw"'],
      dtype='object', name='book_name', length=706)

In [34]:
df = filter2[filter2].index
df

Index(['1984', '1st to Die: A Novel', '2nd Chance', '4 Blondes',
       'A Bend in the Road', 'A Case of Need',
       'A Child Called \It\": One Child's Courage to Survive"',
       'A Civil Action', 'A Day Late and a Dollar Short', 'A Fine Balance',
       ...
       'Winter Solstice', 'Wish You Well', 'Without Remorse',
       'Wizard and Glass (The Dark Tower, Book 4)', 'Wuthering Heights',
       'Year of Wonders', 'You Belong To Me',
       'Zen and the Art of Motorcycle Maintenance: An Inquiry into Values',
       'Zoya', '\O\" Is for Outlaw"'],
      dtype='object', name='book_name', length=706)

In [35]:
df = filter1[filter1["book_name"].isin(df)]
df

user_id        isbn  ratings  \
1150      277427  002542730X       10   
1163      277427  0060930535        0   
1165      277427  0060934417        0   
1168      277427  0061009059        9   
1174      277427  006440188X        0   
...          ...         ...      ...   
1029188   275970  1400031354        0   
1029189   275970  1400031362        0   
1029262   275970  1573229725        0   
1029301   275970  1586210661        9   
1029302   275970  1586212230        0   

                                                 book_name  \
1150     Politically Correct Bedtime Stories: Modern Ta...   
1163                         The Poisonwood Bible: A Novel   
1165                                    Bel Canto: A Novel   
1168     One for the Money (Stephanie Plum Novels (Pape...   
1174                                     The Secret Garden   
...                                                    ...   
1029188  Tears of the Giraffe (No.1 Ladies Detective Ag...   
1029189  Morality for Beautiful Girls (No.1 Ladies Dete...   
1029262                                        Fingersmith   
1029301                             Me Talk Pretty One Day   
1029302                                              Naked   

                          author                  publisher  \
1150           James Finn Garner  John Wiley &amp; Sons Inc   
1163          Barbara Kingsolver                  Perennial   
1165                Ann Patchett                  Perennial   
1168             Janet Evanovich                HarperTorch   
1174     Frances Hodgson Burnett               HarperTrophy   
...                          ...                        ...   
1029188   Alexander McCall Smith                     Anchor   
1029189   Alexander McCall Smith                     Anchor   
1029262             Sarah Waters            Riverhead Books   
1029301            David Sedaris    Time Warner Audio Major   
1029302            David Sedaris    Time Warner Audio Major   

                                                 image_url  
1150     http://images.amazon.com/images/P/002542730X.0...  
1163     http://images.amazon.com/images/P/0060930535.0...  
1165     http://images.amazon.com/images/P/0060934417.0...  
1168     http://images.amazon.com/images/P/0061009059.0...  
1174     http://images.amazon.com/images/P/006440188X.0...  
...                                                    ...  
1029188  http://images.amazon.com/images/P/1400031354.0...  
1029189  http://images.amazon.com/images/P/1400031362.0...  
1029262  http://images.amazon.com/images/P/1573229725.0...  
1029301  http://images.amazon.com/images/P/1586210661.0...  
1029302  http://images.amazon.com/images/P/1586212230.0...  

[58586 rows x 7 columns]

In [36]:
pt = df.pivot_table(
    index="book_name",
    columns="user_id",
    values="ratings"
    )
pt

user_id                                             254     2276    2766    \
book_name                                                                    
1984                                                   9.0     NaN     NaN   
1st to Die: A Novel                                    NaN     NaN     NaN   
2nd Chance                                             NaN    10.0     NaN   
4 Blondes                                              NaN     NaN     NaN   
A Bend in the Road                                     0.0     NaN     7.0   
...                                                    ...     ...     ...   
Year of Wonders                                        NaN     NaN     NaN   
You Belong To Me                                       NaN     NaN     NaN   
Zen and the Art of Motorcycle Maintenance: An I...     NaN     NaN     NaN   
Zoya                                                   NaN     NaN     NaN   
\O\" Is for Outlaw"                                    NaN     NaN     NaN   

user_id                                             2977    3363    4017    \
book_name                                                                    
1984                                                   NaN     NaN     NaN   
1st to Die: A Novel                                    NaN     NaN     NaN   
2nd Chance                                             NaN     NaN     NaN   
4 Blondes                                              NaN     NaN     NaN   
A Bend in the Road                                     NaN     NaN     NaN   
...                                                    ...     ...     ...   
Year of Wonders                                        7.0     NaN     NaN   
You Belong To Me                                       NaN     NaN     NaN   
Zen and the Art of Motorcycle Maintenance: An I...     NaN     0.0     NaN   
Zoya                                                   NaN     NaN     NaN   
\O\" Is for Outlaw"                                    NaN     NaN     NaN   

user_id                                             4385    6251    6323    \
book_name                                                                    
1984                                                   NaN     NaN     NaN   
1st to Die: A Novel                                    NaN     NaN     NaN   
2nd Chance                                             NaN     NaN     NaN   
4 Blondes                                              NaN     0.0     NaN   
A Bend in the Road                                     NaN     NaN     NaN   
...                                                    ...     ...     ...   
Year of Wonders                                        NaN     NaN     NaN   
You Belong To Me                                       NaN     NaN     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     NaN     0.0     NaN   
Zoya                                                   NaN     NaN     NaN   
\O\" Is for Outlaw"                                    NaN     NaN     NaN   

user_id                                             6543    ...  271705  \
book_name                                                   ...           
1984                                                   NaN  ...    10.0   
1st to Die: A Novel                                    9.0  ...     NaN   
2nd Chance                                             0.0  ...     NaN   
4 Blondes                                              NaN  ...     NaN   
A Bend in the Road                                     NaN  ...     NaN   
...                                                    ...  ...     ...   
Year of Wonders                                        0.0  ...     NaN   
You Belong To Me                                       NaN  ...     NaN   
Zen and the Art of Motorcycle Maintenance: An I...     NaN  ...     NaN   
Zoya                                                   NaN  ...     NaN   
\O\" Is for Outlaw"                                    

In [37]:
pt.fillna(0, inplace=True)

In [38]:
pt

user_id                                             254     2276    2766    \
book_name                                                                    
1984                                                   9.0     0.0     0.0   
1st to Die: A Novel                                    0.0     0.0     0.0   
2nd Chance                                             0.0    10.0     0.0   
4 Blondes                                              0.0     0.0     0.0   
A Bend in the Road                                     0.0     0.0     7.0   
...                                                    ...     ...     ...   
Year of Wonders                                        0.0     0.0     0.0   
You Belong To Me                                       0.0     0.0     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     0.0     0.0     0.0   
Zoya                                                   0.0     0.0     0.0   
\O\" Is for Outlaw"                                    0.0     0.0     0.0   

user_id                                             2977    3363    4017    \
book_name                                                                    
1984                                                   0.0     0.0     0.0   
1st to Die: A Novel                                    0.0     0.0     0.0   
2nd Chance                                             0.0     0.0     0.0   
4 Blondes                                              0.0     0.0     0.0   
A Bend in the Road                                     0.0     0.0     0.0   
...                                                    ...     ...     ...   
Year of Wonders                                        7.0     0.0     0.0   
You Belong To Me                                       0.0     0.0     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     0.0     0.0     0.0   
Zoya                                                   0.0     0.0     0.0   
\O\" Is for Outlaw"                                    0.0     0.0     0.0   

user_id                                             4385    6251    6323    \
book_name                                                                    
1984                                                   0.0     0.0     0.0   
1st to Die: A Novel                                    0.0     0.0     0.0   
2nd Chance                                             0.0     0.0     0.0   
4 Blondes                                              0.0     0.0     0.0   
A Bend in the Road                                     0.0     0.0     0.0   
...                                                    ...     ...     ...   
Year of Wonders                                        0.0     0.0     0.0   
You Belong To Me                                       0.0     0.0     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     0.0     0.0     0.0   
Zoya                                                   0.0     0.0     0.0   
\O\" Is for Outlaw"                                    0.0     0.0     0.0   

user_id                                             6543    ...  271705  \
book_name                                                   ...           
1984                                                   0.0  ...    10.0   
1st to Die: A Novel                                    9.0  ...     0.0   
2nd Chance                                             0.0  ...     0.0   
4 Blondes                                              0.0  ...     0.0   
A Bend in the Road                                     0.0  ...     0.0   
...                                                    ...  ...     ...   
Year of Wonders                                        0.0  ...     0.0   
You Belong To Me                                       0.0  ...     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     0.0  ...     0.0   
Zoya                                                   0.0  ...     0.0   
\O\" Is for Outlaw"                                    

In [39]:
from sklearn.metrics.pairwise import cosine_similarity
similarity = cosine_similarity(pt)

In [40]:
similarity[0]

array([1.        , 0.10255025, 0.01220856, 0.        , 0.05367224,
       0.02774901, 0.08216491, 0.13732869, 0.03261686, 0.03667591,
       0.02322418, 0.06766487, 0.02083978, 0.09673735, 0.13388865,
       0.08303112, 0.11153543, 0.05100411, 0.02517784, 0.11706383,
       0.        , 0.14333793, 0.07847534, 0.06150451, 0.08723968,
       0.        , 0.07009814, 0.13658681, 0.07600328, 0.12167134,
       0.00768046, 0.01473221, 0.        , 0.07965814, 0.04522617,
       0.01556271, 0.09495938, 0.0182307 , 0.02610465, 0.07984012,
       0.11679969, 0.0569124 , 0.08354155, 0.08471898, 0.08785938,
       0.05491435, 0.0548505 , 0.27026514, 0.09779123, 0.06016046,
       0.08958835, 0.06748675, 0.        , 0.04468098, 0.01920872,
       0.        , 0.05629067, 0.00557964, 0.07877059, 0.05219479,
       0.18908177, 0.        , 0.01240656, 0.02984572, 0.04279502,
       0.12680125, 0.16566735, 0.        , 0.13357242, 0.06615478,
       0.        , 0.        , 0.        , 0.10968075, 0.02806

#### Fetch the index of the book

In [41]:
pt.index == "Year of Wonders"

array([False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False,

In [42]:
np.where(pt.index == "Year of Wonders")

(array([701]),)

In [43]:
np.where(pt.index == "Year of Wonders")[0]

array([701])

In [44]:
np.where(pt.index == "Year of Wonders")[0][0]

np.int64(701)

In [45]:
def dummy1(book_name):
    # fetch index of the book_name
    book_index = np.where(pt.index == book_name)[0][0]
    return book_index

In [46]:
dummy1("2nd Chance")

np.int64(2)

#### Fetch similar ites of the book

In [47]:
similarity[0]

array([1.        , 0.10255025, 0.01220856, 0.        , 0.05367224,
       0.02774901, 0.08216491, 0.13732869, 0.03261686, 0.03667591,
       0.02322418, 0.06766487, 0.02083978, 0.09673735, 0.13388865,
       0.08303112, 0.11153543, 0.05100411, 0.02517784, 0.11706383,
       0.        , 0.14333793, 0.07847534, 0.06150451, 0.08723968,
       0.        , 0.07009814, 0.13658681, 0.07600328, 0.12167134,
       0.00768046, 0.01473221, 0.        , 0.07965814, 0.04522617,
       0.01556271, 0.09495938, 0.0182307 , 0.02610465, 0.07984012,
       0.11679969, 0.0569124 , 0.08354155, 0.08471898, 0.08785938,
       0.05491435, 0.0548505 , 0.27026514, 0.09779123, 0.06016046,
       0.08958835, 0.06748675, 0.        , 0.04468098, 0.01920872,
       0.        , 0.05629067, 0.00557964, 0.07877059, 0.05219479,
       0.18908177, 0.        , 0.01240656, 0.02984572, 0.04279502,
       0.12680125, 0.16566735, 0.        , 0.13357242, 0.06615478,
       0.        , 0.        , 0.        , 0.10968075, 0.02806

In [48]:
enumerate(similarity[0])

In [49]:
list(enumerate(similarity[0]))

[(0, np.float64(0.9999999999999998)),
 (1, np.float64(0.10255024829874466)),
 (2, np.float64(0.012208555646509366)),
 (3, np.float64(0.0)),
 (4, np.float64(0.053672244225442485)),
 (5, np.float64(0.027749013024662858)),
 (6, np.float64(0.08216491130000354)),
 (7, np.float64(0.13732869060263378)),
 (8, np.float64(0.03261686350494277)),
 (9, np.float64(0.03667590746512177)),
 (10, np.float64(0.02322418425912186)),
 (11, np.float64(0.06766486753662343)),
 (12, np.float64(0.020839782065795794)),
 (13, np.float64(0.09673734976610586)),
 (14, np.float64(0.13388864701187034)),
 (15, np.float64(0.0830311169181103)),
 (16, np.float64(0.1115354275858341)),
 (17, np.float64(0.05100410510067514)),
 (18, np.float64(0.025177841632809915)),
 (19, np.float64(0.11706383185253948)),
 (20, np.float64(0.0)),
 (21, np.float64(0.14333792815336935)),
 (22, np.float64(0.07847534327232125)),
 (23, np.float64(0.06150451339525138)),
 (24, np.float64(0.08723968326958673)),
 (25, np.float64(0.0)),
 (26, np.float64

In [50]:
sorted(list(enumerate(similarity[0])))

[(0, np.float64(0.9999999999999998)),
 (1, np.float64(0.10255024829874466)),
 (2, np.float64(0.012208555646509366)),
 (3, np.float64(0.0)),
 (4, np.float64(0.053672244225442485)),
 (5, np.float64(0.027749013024662858)),
 (6, np.float64(0.08216491130000354)),
 (7, np.float64(0.13732869060263378)),
 (8, np.float64(0.03261686350494277)),
 (9, np.float64(0.03667590746512177)),
 (10, np.float64(0.02322418425912186)),
 (11, np.float64(0.06766486753662343)),
 (12, np.float64(0.020839782065795794)),
 (13, np.float64(0.09673734976610586)),
 (14, np.float64(0.13388864701187034)),
 (15, np.float64(0.0830311169181103)),
 (16, np.float64(0.1115354275858341)),
 (17, np.float64(0.05100410510067514)),
 (18, np.float64(0.025177841632809915)),
 (19, np.float64(0.11706383185253948)),
 (20, np.float64(0.0)),
 (21, np.float64(0.14333792815336935)),
 (22, np.float64(0.07847534327232125)),
 (23, np.float64(0.06150451339525138)),
 (24, np.float64(0.08723968326958673)),
 (25, np.float64(0.0)),
 (26, np.float64

In [51]:
sorted(list(enumerate(similarity[0])), key=lambda x: x[1], reverse=True)

[(0, np.float64(0.9999999999999998)),
 (47, np.float64(0.2702651417103732)),
 (545, np.float64(0.2639619371123497)),
 (82, np.float64(0.23669374347400993)),
 (634, np.float64(0.23299389358170394)),
 (551, np.float64(0.2262639743141286)),
 (181, np.float64(0.22222254415660225)),
 (500, np.float64(0.21057743765309012)),
 (344, np.float64(0.20675540390453995)),
 (554, np.float64(0.20025063706783053)),
 (520, np.float64(0.19972624710300058)),
 (527, np.float64(0.19656287691544358)),
 (488, np.float64(0.19511819825564608)),
 (552, np.float64(0.1920889620576415)),
 (318, np.float64(0.19182230688281934)),
 (624, np.float64(0.1892531370760997)),
 (598, np.float64(0.189213677823462)),
 (60, np.float64(0.18908176672188648)),
 (622, np.float64(0.18571699439818884)),
 (647, np.float64(0.1841081817659337)),
 (546, np.float64(0.18408910419891056)),
 (202, np.float64(0.18049207281616433)),
 (541, np.float64(0.17990898085771034)),
 (309, np.float64(0.17415730924332484)),
 (588, np.float64(0.1738182944

In [52]:
sorted(list(enumerate(similarity[0])), key=lambda x: x[1], reverse=True)[1:6]

[(47, np.float64(0.2702651417103732)),
 (545, np.float64(0.2639619371123497)),
 (82, np.float64(0.23669374347400993)),
 (634, np.float64(0.23299389358170394)),
 (551, np.float64(0.2262639743141286))]

In [53]:
def dummy2(book_name):
    # fetch index of the book_name
    book_index = np.where(pt.index == book_name)[0][0]
    simmilar_items = sorted(list(enumerate(similarity[book_index])), key=lambda x: x[1], reverse=True)[1:6]
    
    lists = []
    for item in simmilar_items:
        print(item[0])

In [54]:
dummy2("2nd Chance")

203
577
666
575
198


#### Fetch name of the recommend book

In [55]:
pt.index

Index(['1984', '1st to Die: A Novel', '2nd Chance', '4 Blondes',
       'A Bend in the Road', 'A Case of Need',
       'A Child Called \It\": One Child's Courage to Survive"',
       'A Civil Action', 'A Day Late and a Dollar Short', 'A Fine Balance',
       ...
       'Winter Solstice', 'Wish You Well', 'Without Remorse',
       'Wizard and Glass (The Dark Tower, Book 4)', 'Wuthering Heights',
       'Year of Wonders', 'You Belong To Me',
       'Zen and the Art of Motorcycle Maintenance: An Inquiry into Values',
       'Zoya', '\O\" Is for Outlaw"'],
      dtype='object', name='book_name', length=706)

In [56]:
pt.index[47]

'Animal Farm'

In [57]:
def recommend(book_name):
    # fetch index of the book_name
    book_index = np.where(pt.index == book_name)[0][0]
    simmilar_items = sorted(list(enumerate(similarity[book_index])), key=lambda x: x[1], reverse=True)[1:6]
    
    recommendations = []
    
    for item in simmilar_items:
        recommendations.append(pt.index[item[0]])
    return recommendations

In [58]:
recommend("The Notebook")

['A Walk to Remember',
 'The Rescue',
 'One Door Away from Heaven',
 'Toxin',
 'The Five People You Meet in Heaven']

In [59]:
model_data = {
    "data": books,
    "similarity": similarity,
     "pivot_index": pt.index.tolist()  # <-- save only books used for similarity
}

In [60]:
import joblib
joblib.dump(model_data, "model.pkl")
print("Model Saved Successfully!")

Model Saved Successfully!


In [61]:
loaded_model = joblib.load("model.pkl")
loaded_model_prediction = recommend("Toxin")
print(f"Prediction from loaded model: {loaded_model_prediction}")

for book in loaded_model_prediction:
    print(book)

Prediction from loaded model: ['Dolores Claiborne', 'Invasion', 'Watchers', 'Acceptable Risk', 'Vector']
Dolores Claiborne
Invasion
Watchers
Acceptable Risk
Vector
